/*
 * Copyright(c) 2012-2018 Intel Corporation
 * SPDX-License-Identifier: BSD-3-Clause-Clear
 */

syntax = "proto3";
option cc_generic_services = true;
import "defs.proto";
import "opts.proto";
import "statistics.proto";
package octf.proto;

message ParseTraceRequest {
    string tracePath = 1 [
        (opts_param).cli_required = true,
        (opts_param).cli_long_key = "path",
        (opts_param).cli_short_key = "p",
        (opts_param).cli_desc = "Path to trace"
    ];

    OutputFormat format = 2;

    bool raw = 3 [
        (opts_param).cli_required = false,
        (opts_param).cli_long_key = "raw",
        (opts_param).cli_short_key = "r",
        (opts_param).cli_desc =
            "Present trace as it had been recorded without post processing"
    ];
}

enum ParseTraceTimeUnit {
    option (opts_enum_param).cli_enum.default_value = 0;
    option (opts_enum_param).cli_required = false;
    option (opts_enum_param).cli_short_key = "t";
    option (opts_enum_param).cli_long_key = "time-unit";
    option (opts_enum_param).cli_desc = "Specifiy time unit of parsing result";

    ns = 0 [
        (opts_enumval).cli_desc = "nanoseconds",
        (opts_enumval).cli_switch = "ns"
    ];

    us = 1 [
        (opts_enumval).cli_desc = "microseconds",
        (opts_enumval).cli_switch = "us"
    ];
}

message GetTraceStatisticsRequest {
    string tracePath = 1 [
        (opts_param).cli_required = true,
        (opts_param).cli_long_key = "path",
        (opts_param).cli_short_key = "p",
        (opts_param).cli_desc = "Path to trace"
    ];

    OutputFormat format = 2;

    ParseTraceTimeUnit timeUnit = 3;
}

service InterfaceTraceParsing {
    option (opts_interface).cli = true;

    option (opts_interface).version = 1;

    rpc ParseTrace(ParseTraceRequest) returns (Void) {
        option (opts_command).cli = true;

        option (opts_command).cli_long_key = "parse-trace";

        option (opts_command).cli_short_key = "P";

        option (opts_command).cli_desc =
            "Parse trace files to human readable form";
    }

    rpc GetTraceStatistics(GetTraceStatisticsRequest)
        returns (IoStatisticsSet) {
        option (opts_command).cli = true;

        option (opts_command).cli_long_key = "get-trace-statistics";

        option (opts_command).cli_short_key = "T";

        option (opts_command).cli_desc = "Get trace statistics";
    }
}
